### PR TITLE
feat(apps): create credential service server

### PR DESCRIPTION
**Pull request by josephperrott**
Create the credential service server. When logged in with an ng-dev token, allows for a user to request via websocket, a temporary Github credential.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -113,6 +116,8 @@
     "cli-progress": "^3.7.0",
     "conventional-commits-parser": "^3.2.1",
     "ejs": "^3.1.6",
+    "express": "^4.18.1",
+    "express-ws": "^5.0.2",
     "firebase": "9.9.0",
```
Comment (devversion) [inline package.json:121]: Do we use express actually?

File: apps/credential-service/lib/server.ts [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+}
+
+/**
+ * WebSocket handler to generate temporary Github access token.
+ *
+ * The access token is automatically revoked when the websocket is closed.
+ */
+async function wsHandler(ws: WebSocket, req: IncomingMessage) {
+  /** Whether the websocket heartbeat check is still alive. */
+  let hasHeartbeat: boolean;
```
Comment (devversion) [inline apps/credential-service/lib/server.ts:60]:
```suggestion
  let receivedHeartbeatResponse: boolean;
```

File: apps/credential-service/lib/server.ts [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+  } catch (e) {
+    console.error('Unable to verified authorized user');
+    console.error(e);
+    socket.write('HTTP/1.1 401 Unauthorized\r\n\r\n');
+    socket.destroy();
+    return;
+  }
+
+  /** The repository name and owner for the request. */
+  const {repo, owner} = req.headers as {repo: string; owner: string};
```
Comment (devversion) [inline apps/credential-service/lib/server.ts:127]: can we make the headers more scoped to the custom use-case. Like: NgDevTargetRepoName and so on?

File: apps/credential-service/lib/server.ts [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+
+/** The http web server. */
+const server = createServer();
+/** The websocket server to handle websocket requests. */
+const wss = new WebSocketServer({noServer: true});
+
+wss.on('connection', wsHandler);
+server.on('upgrade', upgradeHandler);
+server.on('listening', () => console.log('Credential Service startup complete, listening'));
+server.listen(parseInt(PORT, 10));
```
Comment (devversion) [inline apps/credential-service/lib/server.ts:149]:
```suggestion
server.listen(Number(PORT));
```

**Review by devversion — COMMENTED** [review]
Should there be a docker file?

File: package.json (hunk 1 above)
Comment (josephperrott) [reply package.json]: Bad rebase caused it to stick around.

**Comment by josephperrott** [comment]
Updated the commit message originally, but then Github decided to not update the title and summary for the PR when I pushed again.  Its corrected now.

File: apps/credential-service/BUILD.bazel [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+
+copy_to_bin(
+    name = "bin_files",
+    srcs = [
+        "package.json",
+    ],
+)
+
+esbuild_esm_bundle(
+    name = "credential-service",
```
Comment (devversion) [inline apps/credential-service/BUILD.bazel:20]: ooc: how do you intend to deploy the script?

File: apps/credential-service/lib/server.ts [hunk 6]
```diff
@@ -0,0 +1,160 @@
...
+  };
+}
+
+/**
+ * WebSocket handler to generate temporary Github access token.
+ *
+ * The access token is automatically revoked when the websocket is closed.
+ */
+async function wsHandler(ws: WebSocket, req: IncomingMessage) {
+  /** Whether the websocket heartbeat check is still alive. */
```
Comment (devversion) [inline apps/credential-service/lib/server.ts:64]: nit: comment is not reflecting what it actually represents

File: apps/credential-service/lib/server.ts [hunk 7]
```diff
@@ -0,0 +1,160 @@
...
+ * Handle upgrade requests before the websocket is used. Enforces authentication mechanisms
+ * and ensuring the required data is present in the request.
+ */
+async function upgradeHandler(req: IncomingMessage, socket: Duplex, head: Buffer) {
+  try {
+    if (!authorizationRegex.test(req.headers.authorization!)) {
+      throw Error('Missing or invalid authorization header syntax');
+    }
+    /** The NgDev token from the user to be verified. */
+    const [_, ngDevToken] = req.headers.authorization!.match(/Bearer (.*)/)!;
```
Comment (devversion) [inline apps/credential-service/lib/server.ts:119]: You have a regex for that above

File: apps/credential-service/lib/server.ts [hunk 8]
```diff
@@ -0,0 +1,160 @@
...
+  ws.send(token);
+}
+
+/**
+ * Handle upgrade requests before the websocket is used. Enforces authentication mechanisms
+ * and ensuring the required data is present in the request.
+ */
+async function upgradeHandler(req: IncomingMessage, socket: Duplex, head: Buffer) {
+  try {
+    if (!authorizationRegex.test(req.headers.authorization!)) {
```
Comment (devversion) [inline apps/credential-service/lib/server.ts:115]: Super nit (while I leave actual comments): Is the exclamation needed? generally a good thing to avoid

**Review by devversion — APPROVED** [review]
LGTM, just some minor comments

File: apps/credential-service/BUILD.bazel (hunk 5 above)
Comment (josephperrott) [reply apps/credential-service/BUILD.bazel:20]: My plan is to handle it in a follow up, the plan is to run the service in a container.  I was looking at `rules_docker` but its set to be replaced so I am still sorting out specifics.

File: apps/credential-service/lib/server.ts (hunk 8 above)
Comment (josephperrott) [reply apps/credential-service/lib/server.ts:115]: Switched it to a `|| ''`, which ends up functioning the same for `Regex.test`

File: apps/credential-service/BUILD.bazel (hunk 5 above)
Comment (devversion) [reply apps/credential-service/BUILD.bazel:20]: Sounds good. was mostly asking out of curiosity given the checked in script

File: apps/credential-service/lib/server.ts [hunk 9]
```diff
@@ -112,11 +112,14 @@ async function wsHandler(ws: WebSocket, req: IncomingMessage) {
...
+    if (!authorizationRegex.test(req.headers.authorization || '')) {
       throw Error('Missing or invalid authorization header syntax');
     }
-    /** The NgDev token from the user to be verified. */
-    const [_, ngDevToken] = req.headers.authorization!.match(/Bearer (.*)/)!;
+    /**
+     * The NgDev token from the user to be verified.
+     * We use a non-null assertion as the regex was already tested above.
+     */
+    const [_, ngDevToken] = authorizationRegex.exec(req.headers.authorization!)!;
```
Comment (devversion) [inline apps/credential-service/lib/server.ts:122]: just FYI the whole non-nullability could be trivially inferred by TypeScript if you made an explicit check above

**Review by devversion — APPROVED** [review]
LGTM!

File: apps/credential-service/lib/server.ts (hunk 9 above)
Comment (devversion) [reply apps/credential-service/lib/server.ts:122]: Also lint is failing

**Comment by josephperrott** [comment]
This PR was merged into the repository by commit 1609543e7d76ccf66fbbab1bfa8ae4c2767adb80.

**Pull request closed without merging**